python3: Re-enable ensurepip module

garar opened the pull request:
Signed-off-by: Krystian Kichewko krystiankichewko@gmail.com

Maintainer: @jefferyto @commodo @neheb 
Compile tested: arch64, Raspberry Pi Compute Module 4 IoT Router Carrier Board Mini, OpenWRT 21.02.3
Run tested: arch64, Raspberry Pi Compute Module 4 IoT Router Carrier Board Mini, OpenWRT 21.02.3

Description:

This pull request re-enables ensurepip module that is needed by some software like pipx and pipenv and also this should fix: https://github.com/openwrt/packages/issues/12707

commodo commented:
i'd probably need to check this;

i thought/assumed that selecting `python3-pip` worked fine and it's sufficient;

garar commented:
Hi @commodo !

No worries, no hurry.

I thought exact same thing, but then I installed `python3-pip` and the error was still happening.

When I get back home from work I can get some logs to show what I did.

commodo commented:
> Hi @commodo !
> 
> No worries, no hurry.
> 
> I thought exact same thing, but then I installed `python3-pip` and the error was still happening.
> 
> When I get back home from work I can get some logs to show what I did.

So, I am getting this error.
```
if test "x" != "x" ; then \
	rm -f /home/aardelean/work/openwrt/cmp/build_dir/target-i386_pentium4_musl/Python-3.10.7/ipkg-install/usr/bin/python3-32; \
	(cd /home/aardelean/work/openwrt/cmp/build_dir/target-i386_pentium4_musl/Python-3.10.7/ipkg-install/usr/bin; ln -s python3.10-32 python3-32) \
fi
if test "x" != "x" ; then \
	rm -f /home/aardelean/work/openwrt/cmp/build_dir/target-i386_pentium4_musl/Python-3.10.7/ipkg-install/usr/bin/python3-intel64; \
	(cd /home/aardelean/work/openwrt/cmp/build_dir/target-i386_pentium4_musl/Python-3.10.7/ipkg-install/usr/bin; ln -s python3.10-intel64 python3-intel64) \
fi
rm -f /home/aardelean/work/openwrt/cmp/build_dir/target-i386_pentium4_musl/Python-3.10.7/ipkg-install/usr/man/man1/python3.1
(cd /home/aardelean/work/openwrt/cmp/build_dir/target-i386_pentium4_musl/Python-3.10.7/ipkg-install/usr/man/man1; ln -s python3.10.1 python3.1)
if test "xupgrade" != "xno"  ; then \
	case upgrade in \
		upgrade) ensurepip="--upgrade" ;; \
		install|*) ensurepip="" ;; \
	esac; \
	 _PYTHON_PROJECT_BASE=/home/aardelean/work/openwrt/cmp/build_dir/target-i386_pentium4_musl/Python-3.10.7 _PYTHON_HOST_PLATFORM=linux-i486 PYTHONPATH=/home/aardelean/work/openwrt/cmp/build_dir/target-i386_pentium4_musl/Python-3.10.7/build/lib.linux-i486-3.10:./Lib _PYTHON_SYSCONFIGDATA_NAME=_sysconfigdata python3.10 -m ensurepip \
		$ensurepip --root=/home/aardelean/work/openwrt/cmp/build_dir/target-i386_pentium4_musl/Python-3.10.7/ipkg-install/ ; \
fi
Traceback (most recent call last):
  File "/home/aardelean/work/openwrt/cmp/build_dir/target-i386_pentium4_musl/Python-3.10.7/Lib/subprocess.py", line 69, in <module>
    import msvcrt
ModuleNotFoundError: No module named 'msvcrt'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/aardelean/work/openwrt/cmp/build_dir/target-i386_pentium4_musl/Python-3.10.7/Lib/runpy.py", line 187, in _run_module_as_main
    mod_name, mod_spec, code = _get_module_details(mod_name, _Error)
  File "/home/aardelean/work/openwrt/cmp/build_dir/target-i386_pentium4_musl/Python-3.10.7/Lib/runpy.py", line 146, in _get_module_details
    return _get_module_details(pkg_main_name, error)
  File "/home/aardelean/work/openwrt/cmp/build_dir/target-i386_pentium4_musl/Python-3.10.7/Lib/runpy.py", line 110, in _get_module_details
    __import__(pkg_name)
  File "/home/aardelean/work/openwrt/cmp/build_dir/target-i386_pentium4_musl/Python-3.10.7/Lib/ensurepip/__init__.py", line 4, in <module>
    import subprocess
  File "/home/aardelean/work/openwrt/cmp/build_dir/target-i386_pentium4_musl/Python-3.10.7/Lib/subprocess.py", line 74, in <module>
    import _posixsubprocess
ImportError: /home/aardelean/work/openwrt/cmp/build_dir/target-i386_pentium4_musl/Python-3.10.7/build/lib.linux-i486-3.10/_posixsubprocess.cpython-310.so: wrong ELF class: ELFCLASS32
make[4]: *** [Makefile:1280: install] Error 1
make[4]: Leaving directory '/home/aardelean/work/openwrt/cmp/build_dir/target-i386_pentium4_musl/Python-3.10.7'
make[3]: *** [Makefile:386: /home/aardelean/work/openwrt/cmp/build_dir/target-i386_pentium4_musl/Python-3.10.7/.built] Error 2
make[3]: Leaving directory '/home/aardelean/work/openwrt/cmp/feeds/packages/lang/python/python3'
time: package/feeds/packages/python3/compile#747.11#29.20#252.10
    ERROR: package/feeds/packages/python3 failed to build.
make[2]: *** [package/Makefile:116: package/feeds/packages/python3/compile] Error 1
make[2]: Leaving directory '/home/aardelean/work/openwrt/cmp'
make[1]: *** [package/Makefile:110: /home/aardelean/work/openwrt/cmp/staging_dir/target-i386_pentium4_musl/stamp/.package_compile] Error 2
make[1]: Leaving directory '/home/aardelean/work/openwrt/cmp'
make: *** [/home/aardelean/work/openwrt/cmp/include/toplevel.mk:231: world] Error 2
```

i'm wondering if you are compiling for aarch64 on an aarch64, and the issue may not be visible?

let me try something, in the meantime;

garar commented:
Hi!

Logs from my system before any patches, to show that `python3-pip` is not enough:

```
[25/01/23 | 13:19]  ~ opkg update
Downloading https://downloads.openwrt.org/releases/21.02.3/targets/bcm27xx/bcm2711/packages/Packages.gz
Updated list of available packages in /var/opkg-lists/openwrt_core
Downloading https://downloads.openwrt.org/releases/21.02.3/targets/bcm27xx/bcm2711/packages/Packages.sig
Signature check passed.
Downloading https://downloads.openwrt.org/releases/21.02.3/packages/aarch64_cortex-a72/base/Packages.gz
Updated list of available packages in /var/opkg-lists/openwrt_base
Downloading https://downloads.openwrt.org/releases/21.02.3/packages/aarch64_cortex-a72/base/Packages.sig
Signature check passed.
Downloading https://downloads.openwrt.org/releases/21.02.3/packages/aarch64_cortex-a72/luci/Packages.gz
Updated list of available packages in /var/opkg-lists/openwrt_luci
Downloading https://downloads.openwrt.org/releases/21.02.3/packages/aarch64_cortex-a72/luci/Packages.sig
Signature check passed.
Downloading https://downloads.openwrt.org/releases/21.02.3/packages/aarch64_cortex-a72/packages/Packages.gz
Updated list of available packages in /var/opkg-lists/openwrt_packages
Downloading https://downloads.openwrt.org/releases/21.02.3/packages/aarch64_cortex-a72/packages/Packages.sig
Signature check passed.
Downloading https://downloads.openwrt.org/releases/21.02.3/packages/aarch64_cortex-a72/routing/Packages.gz
Updated list of available packages in /var/opkg-lists/openwrt_routing
Downloading https://downloads.openwrt.org/releases/21.02.3/packages/aarch64_cortex-a72/routing/Packages.sig
Signature check passed.
Downloading https://downloads.openwrt.org/releases/21.02.3/packages/aarch64_cortex-a72/telephony/Packages.gz
Updated list of available packages in /var/opkg-lists/openwrt_telephony
Downloading https://downloads.openwrt.org/releases/21.02.3/packages/aarch64_cortex-a72/telephony/Packages.sig
Signature check passed.
Time: 0h:00m:03s                                                                                                                                                                                                                                              
[25/01/23 | 13:19]  ~ opkg install python3 python3-pip
Package python3 (3.9.16-1) installed in root is up to date.
Package python3-pip (22.0.4-1) installed in root is up to date.
[25/01/23 | 13:19]  ~ pipx install pyenv
Error: Command '['/root/.local/pipx/shared/bin/python3', '-Im', 'ensurepip', '--upgrade', '--default-pip']' returned non-zero exit status 1.

'/usr/bin/python3 -m venv --clear /root/.local/pipx/shared' failed
[25/01/23 | 13:21]  ~ /root/.local/pipx/shared/bin/python3 -Im ensurepip --upgrade --default-pip
/root/.local/pipx/shared/bin/python3: No module named ensurepip
[25/01/23 | 13:21]  ~ 
[25/01/23 | 13:21]  ~ python3 -Im ensurepip --upgrade --default-pip 
/usr/bin/python3: No module named ensurepip
[25/01/23 | 13:25] ~
```

About your problems, I don't know openwrt build system enough, but why is your logs say Python 3.10.7?
Also, are there any docs on how to properly check my MR? I've read CONTRIBUTING.md and https://openwrt.org/docs/guide-developer/packages and I might be doing something wrong, and this is why it is working for me.

Thanks for checking it!

commodo commented:
I got good results with this https://github.com/openwrt/packages/pull/20387

the unfortunate result is that it increases the python3-pip package from 6 to 9+ MB
but, virtualenv is an interesting idea for OpenWrt

garar commented:
Ok, so if https://github.com/openwrt/packages/pull/20387 looks better, I'm closing my pull request.
